rfcs/text/0030/process.yml

```diff
@@ -0,0 +1,310 @@
+      description: >
+        Thread name.
+
+    - name: start
+      level: extended
+      type: date
+      example: "2016-05-23T08:05:34.853Z"
+      description: >
+        The time this process started.
+
+    - name: ppid_start
+      level: extended
+      type: date
+      example: "2016-05-23T08:05:34.853Z"
+      description: >
+        The time this process's parent started.
+
+    - name: pgid_start
+      level: extended
+      type: date
+      example: "2016-05-23T08:05:34.853Z"
+      description: >
+        The time this process's process group started.
+
+    - name: sid_start
+      level: extended
+      type: date
+      example: "2016-05-23T08:05:34.853Z"
+      description: >
+        The time this process's session started.
+
+    - name: uptime
+      level: extended
+      type: long
+      example: 1325
+      description: >
+        Seconds the process has been up.
+
+    - name: working_directory
+      level: extended
+      type: keyword
+      example: /home/alice
+      description: >
+        The working directory of the process.
+      multi_fields:
+        - type: match_only_text
+          name: text
+
+    - name: exit_code
+      level: extended
+      type: long
+      example: 137
+      short: The exit code of the process.
+      description: >
+        The exit code of the process, if this is a termination event.
+
+        The field should be absent if there is no exit code for the event (e.g.
+        process start).
+
+    - name: end
+      level: extended
+      type: date
+      example: "2016-05-23T08:05:34.853Z"
+      description: >
+        The time the process ended.
+
+    - name: interactive
+      level: extended
+      type: boolean
+      example: true
+      short: Whether the process is connected to an interactive shell.
+      description: >
+        Whether the process is connected to an interactive shell.
+
+        Process interactivity is infered from the processes file descriptors. If the character device for the controlling tty is the same as stdin and stderr for the process, the process is considered interactive.
+
+        Note: A non-interactive process can belong to an interactive session. Backgrounded processes are an example of this.
+
+    - name: tty
+      level: extended
+      type: fd
+      example: >
+        process.tty = {
+          char_device: {
+            major: 4,
+            minor: 1
+          }
+        }
+      short: The file descriptor for the controlling tty (ctty).
+      description: >
+        The file descriptor for the controlling tty (ctty) of the session.
+
+        If set, session is interactive.
+
+    - name: env_vars
+      level: extended
+      type: object
+      short: Array of environment variables.
+      description: >
+        Array of environment variables.
+
+        May be filtered to protect sensitive information.
+      example: "[{ key: 'PATH', value: '/usr/bin;/usr/local/bin;'}, { key: 'USER', value: 'ubuntu' }]"
```

Comment:
@dcode @devonakerr @joe-desimone @ricardoungureanu @jrmolin We've discussed adding environment variables in the past to enable EQL (on Endpoint or in SIEM) and Detection Engine rules so I'm tagging you here.

Does anyone have thoughts on how to structure environment variables in ECS? I think there are two options, each with costs and benefits. Does anyone feel strongly one way or the other? Is there a third option I didn't think of?

**Option 1: An array**
This is what @mitodrummer suggested.

Example:
```json
{
    "env_vars": [
        { "key": "PATH", "value": "/usr/bin;/usr/local/bin;"},
        { "key": "USER", "value": "ubuntu" }
    ]
}
```

This is very flexible. It allows new environment variables to be added by any agent, even by runtime configuration and then allows those fields to be indexed in Elasticsearch since existing index mappings would not need to change.

A drawback is that consumers of this data (e.g. EQL or Detection Engine) need to iterate over an array when executing rules to see if key environment variables exist and contain key values. Also, the `env_vars` would need to be indexed in a nested way to be usable for queries, though I don't know what performance or usability impact that might have.

**Option 2: Restricted, enumerated values**

Example:
```json
{
    "env_vars": {
        "PATH": "/usr/bin;/usr/local/bin;",
        "USER": "ubuntu"
    }
}
```

The cost/benefits are basically the exact opposite of option 1.

A benefit of this approach is it is easy to access and use environment variables in a performant way. The drawback is that causes a heavyweight process to add new variables. ECS agents would need to RFC new environment variables into ECS before they can be collected.

Reply:
With the second option we have to ensure that env var naming restrictions align with those JSON object field names.    They appear to be close for Linux and MacOS (POSIX influence).  Not sure if we would apply this to Windows but I don't have their naming constraints at my fingertips.

```Environment variable names used by the utilities in the Shell and Utilities volume of IEEE Std 1003.1-2001 consist solely of uppercase letters, digits, and the '_' (underscore) from the characters defined in Portable Character Set and do not begin with a digit```

Reply:
> Also, the env_vars would need to be indexed in a nested way to be usable for queries, though I don't know what performance or usability impact that might have

Yes, the `nested` data type is necessary anytime the independence of each object in the array needs to be maintained since ES doesn't have a concept of inner objects. Nested fields need `nested` queries and aggs to work as expected, and Kibana has [limited support](https://github.com/elastic/kibana/issues/1084#issuecomment-585178079). And then, on top of querying and visualization limitations, the performance concerns are another consideration.

Reply:
I've just updated process.env_vars to be of type 'object' (not using an array). I've tested this in my local kibana setup, and KQL seems to expand into the object to find it's keys. So I am able to write KQL like:     process.env_vars.HOME: '/home/kg' AND process.env_vars.USER: 'kg'. It seems type 'object' will dynamically add new index mappings? I am fairly new to ES so correct me I'm missing something important wrt that statement.

Reply:
@ferullo You mention that "ECS agents would need to RFC new environment variables into ECS before they can be collected." 

Is that true if we just make env_vars type object? KQL seems to have no problem dynamically mapping new fields inside said object. Will rule engines like EQL be able to make use of that, e.g process WHERE process.env_vars.HOME: '/home/kg' ?

Reply:
TIL! I had no idea `object` mapping would solve that issue, index mappings are definitely outside my area of expertise. If EQL has no trouble with it then I think we're good to use the Option 2 format. Am I wrong that there is general agreement that if Option 2 is preferred if it's possible?

@kevinlog do you know if there are any reasons using `object` to map a dynamic `env_vars` fieldset would be problematic?

@mitodrummer since you have a set up that ingests this type of data into Elasticsearch, can you verify Elasticsearch's EQL implementation will work with Option 2? Below is a screenshot of an EQL query I did in my 7.16 stack using Dev Tools to use as reference. I'm also happy to chat over Slack also.

For Endpoint's EQL implementation, Option 2 should just work. At the very least, since Endpoint is both the event document producer and EQL consumer we could find a way to make sure it works.

<img width="2043" alt="image" src="https://user-images.githubusercontent.com/56368752/146412312-5c092540-c472-4ff7-8f0b-9c0d4c97107a.png">

Reply:
Tested EQL with a dynamically mapped index (using the new process event format). The EQL rule engine was able to detect all the new properties and allow me to make use of them. Seems the engine checks the mappings (or some other means of intropection) and allows you to specify them in the rules. So unless I'm missing something it seems 'type: object' should work fine for rules.